cedar-policy-validator/src/schema.rs

```diff
@@ -207,6 +207,86 @@ impl TryFrom<json_schema::Fragment<RawName>> for ValidatorSchema {
 }
 
 impl ValidatorSchema {
+    /// Returns an iterator over every entity type that can be a principal for any action in this schema
+    pub fn principals(&self) -> impl Iterator<Item = &EntityType> {
+        self.action_ids
+            .values()
+            .flat_map(ValidatorActionId::principals)
+    }
+
+    /// Returns an iterator over every entity type that can be a resource for any action in this schema
+    pub fn resources(&self) -> impl Iterator<Item = &EntityType> {
+        self.action_ids
+            .values()
+            .flat_map(ValidatorActionId::resources)
+    }
+
+    /// Returns an iterator over every entity type that can be a principal for `action` in this schema
+    ///
+    /// # Errors
+    ///
+    /// Returns [`None`] if `action` is not found in the schema
+    pub fn principals_for_action(
+        &self,
+        action: &EntityUID,
+    ) -> Option<impl Iterator<Item = &EntityType>> {
+        self.action_ids
+            .get(action)
+            .map(ValidatorActionId::principals)
+    }
+
+    /// Returns an iterator over every entity type that can be a resource for `action` in this schema
+    ///
+    /// # Errors
+    ///
+    /// Returns [`None`] if `action` is not found in the schema
+    pub fn resources_for_action(
+        &self,
+        action: &EntityUID,
+    ) -> Option<impl Iterator<Item = &EntityType>> {
+        self.action_ids
+            .get(action)
+            .map(ValidatorActionId::resources)
+    }
+
+    /// Returns an iterator over all the entity types that can be a parent of `ty`
+    ///
+    /// # Errors
+    ///
+    /// Returns [`None`] if the `ty` is not found in the schema
+    pub fn parents<'a>(
+        &'a self,
+        ty: &'a EntityType,
+    ) -> Option<impl Iterator<Item = &EntityType> + 'a> {
+        if self.entity_types.contains_key(ty) {
+            Some(self.entity_types.values().filter_map(|ety| {
+                if ety.descendants.contains(ty) {
+                    Some(&ety.name)
+                } else {
+                    None
+                }
+            }))
+        } else {
+            None
+        }
+    }
+
+    /// Returns an iterator over all the action groups defined in this schema
+    pub fn action_groups(&self) -> impl Iterator<Item = &EntityUID> {
+        self.action_ids.values().filter_map(|action| {
+            if action.descendants.is_empty() {
+                None
+            } else {
+                Some(&action.name)
+            }
+        })
+    }
+
+    /// Returns an iterator over all actions defined in this schema
+    pub fn action_euids(&self) -> impl Iterator<Item = &EntityUID> {
```

Comment:
I suggest just `actions()`.  The above function wasn't named `action_group_euids()`, even though it also returns an iterator of `&EntityUID`